Make sure the ends of linspace are correct regardless of the precision

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91625
* #91621

This operation is usually called with small sizes, so the fact that this
adds a couple of operations should be alright. Even more, given the
structure of the data, the branching in the `where` is pretty much free.

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

lezcano committed:
Make sure the ends of linspace are correct regardless of the precision

This operation is usually called with small sizes, so the fact that this
adds a couple of operations should be alright. Even more, given the
structure of the data, the branching in the `where` is pretty much free.

[ghstack-poisoned]

lezcano committed:
Update on "Make sure the ends of linspace are correct regardless of the precision"


This operation is usually called with small sizes, so the fact that this
adds a couple of operations should be alright. Even more, given the
structure of the data, the branching in the `where` is pretty much free.

cc ezyang mruberry ngimel Lezcano peterbell10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -4340,9 +4340,13 @@ def linspace(
-    out = rg * cast((end - start) / (steps - 1)) + cast(start)
+    # We implement torch.lerp without performing rg / (steps - 1) explicitly
+    # With this we get out[0] == start, out[-1] == end
+    step = cast((end - start) / (steps - 1))
```
we shouldn't need to cast `step` to tensor?

lezcano replied:
We need to perform the computation in a higher precision dtype for low-precision dtypes. One way of doing this is by casting `rg` and the other is by casting the numbers to tensors. I chose the latter one, but I could go with the former one if you prefer. I figured that perhaps the latter option only needed to perform manipulations with numbers rather than with the whole array, but the generated code is the same, so I'm fine with either.

lezcano committed:
Update on "Make sure the ends of linspace are correct regardless of the precision"


This operation is usually called with small sizes, so the fact that this
adds a couple of operations should be alright. Even more, given the
structure of the data, the branching in the `where` is pretty much free.

cc ezyang mruberry ngimel Lezcano peterbell10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Make sure the ends of linspace are correct regardless of the precision"


This operation is usually called with small sizes, so the fact that this
adds a couple of operations should be alright. Even more, given the
structure of the data, the branching in the `where` is pretty much free.

cc ezyang mruberry ngimel Lezcano peterbell10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge -f "unrelated errors"
